syft/messaging/plan.py

```diff
@@ -104,111 +95,256 @@ def append(self, key):
-            t = getattr(self.plan, key)
-            if self.is_valid_type(t):
+            obj = getattr(self.plan, key)
+            if isinstance(obj, (FrameworkTensor, FrameworkLayerModule)):
                 self.keys.add(key)
             else:
                 raise ValueError(
-                    f"Obj of type {type(t)} is not supported in the state.\n"
+                    f"Obj of type {type(obj)} is not supported in the state.\n"
                     "Use instead tensors or parameters."
                 )
         return self
 
     def __iadd__(self, other_keys):
         return self.append(other_keys)
 
-    def is_valid_type(self, obj):
-        return isinstance(obj, (torch.nn.Module, torch.Tensor))
-
-    def get_id_at_location(self):
+    def tensors(self) -> List:
         """
-        Returns all the id_at_location of the pointers in the state
+        Fetch and return all the state elements.
+        Perform a check of consistency on the elements ids.
         """
-        id_at_location = []
-
-        for key in self.keys:
-            ptr = getattr(self.plan, key)
-            if isinstance(ptr, torch.nn.Module):
-                for param in ptr.parameters():
-                    id_at_location.append(param.id_at_location)
-            else:
-                id_at_location.append(ptr.id_at_location)
-
-        return id_at_location
+        tensors = []
+        for state_id in self.state_ids:
+            tensor = self.owner.get_obj(state_id)
+            assert tensor.id == state_id
+            tensors.append(tensor)
+        return tensors
+
+    def clone_state(self) -> Dict:
+        """
+        Return a clone of the state elements. Tensor ids are kept.
+        """
+        return {tensor.id: tensor.clone() for tensor in self.tensors()}
 
-    def copy(self):
-        copied_state = {}
-        for key in self.keys:
-            t = getattr(self.plan, key)
-            copied_state[key] = t.copy()
-        return copied_state
+    def copy(self) -> "State":
+        state = State(owner=self.owner, keys=self.keys.copy(), state_ids=self.state_ids.copy())
+        return state
 
     def read(self, key):
         """
-        Returns one element referenced by the store
+        Returns one element referenced by the store of keys.
+        Note that this only works for plans built locally
         """
         return getattr(self.plan, key)
 
-    def set_(self, dict_state):
+    def set_(self, state_dict):
         """
         Reset inplace the state by feeding it a dict of tensors or params
         """
-        for key in self.keys:
-            delattr(self.plan, key)
+        assert list(self.state_ids) == list(state_dict.keys())
 
-        for key in dict_state.keys():
-            setattr(self.plan, key, dict_state[key])
+        for state_id, new_tensor in state_dict.items():
+            tensor = self.owner.get_obj(state_id)
 
-        self.keys = list(dict_state.keys())
+            with torch.no_grad():
```

Comment:
This is torch-specific, and I don't think it should need to be here. Seems like we're in need of a better abstraction if we're going to remain framework agnostic here

Reply:
Any suggestion here?
What would you write if tf to do this?

Reply:
There's not really a tf equivalent for torch.no_grad that I'm aware of. The new GradientTape is used to turn autograd on, which is a clear departure from torch.

It seems like we might be in a situation where Plans require a shared base class that the different frameworks extend. Plans are kind of a first class citizen in TF because of tf.function. This probably requires more than a PR discussion so let's defer it to later

Reply:
Like a generic Tensor which can be implemented by PyTorch & TF ?

Reply:
We kind of already have that in TorchTensor & TensorFlowTensor -- I'm referring to a generic Plan

Reply:
The Plan is already 99% generic
But it makes a few assumptions on Tensors / Modules

Reply:
I'm talking about a 100% generic Plan that torch/tf can both extend as needed to handle their different interfaces. This is in line with how we've handled other core abstractions, e.g. `FrameworkHook`, `FrameworkAttributes`, the pluggable `hook_args`, `NativeTensor`, etc.

Reply:
IMO these core abstractions should be kept as modular as possible, which means deciding on a generic interface and letting implementations extend it as needed

Reply:
I know, I'm working in that direction too :)
But I'm not sure what to restrict to PyTorch, as 99% is already generic

Reply:
This method `set_` seems like a good candidate, as it's called `assign` in TF and would normally be reserved for single variables. In TF this operation would probably be left to the user to perform manually. Another is just below this, `create_grad_if_missing`. Another is referred to in [this comment](https://github.com/OpenMined/PySyft/pull/2627/files#r330910284) -- this is precisely what we'd like to avoid in a generic Plan, since the API should be agnostic to which framework is using/extending it.

In general, it just seems useful to keep in mind that we'll have the following correspondences on our end: `Plan` == `tf.function`, and `Protocol` == `tfe.Protocol`. If we try to maintain parity for both frameworks in a single class, we'll end up handicapping the class, so I think we can avoid that by extending a TorchPlan from this Plan in the future.  It's probably out of scope for this PR, but something worth keeping in mind.

Reply:
I totally agree, however should we do this in this PR or make another one?
Plan were already notoriously not supposed to be in the generic module "as is", so maybe this is a separate topic I would say
@jvmancuso you have been doing amazing job on separating generic from framework specific stuff and you have a vision both on TF & TH, would you be interested to follow up on this?